### ISSUE TITLE
BuildWithSkipAnalyzers is breaking FastUpToDate

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#56883]
This is latest Roslyn main using latest intpreview:

![image](https://user-images.githubusercontent.com/4564579/135519862-045b839e-89e4-4460-a7a2-4d43db60599a.png)

Every build causes this:

```
Build started...
1>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\Microsoft.CodeAnalysis.Collections.Package\Debug\netcoreapp3.1\Microsoft.CodeAnalysis.Collections.Package.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:53 PM) than earliest output 'C:\github\roslyn\artifacts\bin\Microsoft.CodeAnalysis.Collections.Package\Debug\netcoreapp3.1\Microsoft.CodeAnalysis.Collections.Package.dll' (9/30/2021 12:03:35 PM), not up to date. (Microsoft.CodeAnalysis.Collections.Package)
1>------ Build started: Project: Microsoft.CodeAnalysis.Collections.Package, Configuration: Debug Any CPU ------
2>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\Microsoft.CodeAnalysis.XunitHook\Debug\net472\Microsoft.CodeAnalysis.XunitHook.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\bin\Microsoft.CodeAnalysis.XunitHook\Debug\net472\Microsoft.CodeAnalysis.XunitHook.xml' (9/30/2021 12:03:33 PM), not up to date. (Microsoft.CodeAnalysis.XunitHook)
2>------ Build started: Project: Microsoft.CodeAnalysis.XunitHook, Configuration: Debug Any CPU ------
3>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\PrepareTests\Debug\net6.0\PrepareTests.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\obj\PrepareTests\Debug\net6.0\PrepareTests.xml' (9/30/2021 12:03:34 PM), not up to date. (PrepareTests)
3>------ Build started: Project: PrepareTests, Configuration: Debug Any CPU ------
4>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\CompilersIOperationGenerator\Debug\net6.0\CompilersIOperationGenerator.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\bin\CompilersIOperationGenerator\Debug\net6.0\IOperationGenerator.xml' (9/30/2021 12:03:34 PM), not up to date. (CompilersIOperationGenerator)
4>------ Build started: Project: CompilersIOperationGenerator, Configuration: Debug Any CPU ------
5>FastUpToDate: Output 'C:\github\roslyn\artifacts\obj\Roslyn.VisualStudio.Setup.ServiceHub\Debug\net472\Roslyn.VisualStudio.Setup.ServiceHub.dll' does not exist, not up to date. (Roslyn.VisualStudio.Setup.ServiceHub)
5>------ Build started: Project: Roslyn.VisualStudio.Setup.ServiceHub, Configuration: Debug Any CPU ------
6>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\Microsoft.VisualStudio.IntegrationTest.IntegrationService\Debug\net472\Microsoft.VisualStudio.IntegrationTest.IntegrationService.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\bin\Microsoft.VisualStudio.IntegrationTest.IntegrationService\Debug\net472\Microsoft.VisualStudio.IntegrationTest.IntegrationService.xml' (9/30/2021 12:03:33 PM), not up to date. (Microsoft.VisualStudio.IntegrationTest.IntegrationService)
6>------ Build started: Project: Microsoft.VisualStudio.IntegrationTest.IntegrationService, Configuration: Debug Any CPU ------
7>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\Microsoft.CodeAnalysis.PooledObjects.Package\Debug\netstandard1.3\Microsoft.CodeAnalysis.PooledObjects.Package.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\bin\Microsoft.CodeAnalysis.PooledObjects.Package\Debug\netstandard1.3\Microsoft.CodeAnalysis.PooledObjects.Package.dll' (9/30/2021 12:03:34 PM), not up to date. (Microsoft.CodeAnalysis.PooledObjects.Package)
7>------ Build started: Project: Microsoft.CodeAnalysis.PooledObjects.Package, Configuration: Debug Any CPU ------
8>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\Roslyn.VisualStudio.Setup.Dependencies\Debug\net472\Roslyn.VisualStudio.Setup.Dependencies.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\obj\Roslyn.VisualStudio.Setup.Dependencies\Debug\net472\Roslyn.VisualStudio.Setup.Dependencies.xml' (9/30/2021 12:08:18 PM), not up to date. (Roslyn.VisualStudio.Setup.Dependencies)
8>------ Build started: Project: Roslyn.VisualStudio.Setup.Dependencies, Configuration: Debug Any CPU ------
9>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\Microsoft.CodeAnalysis.FunctionResolver\Debug\net45\Microsoft.CodeAnalysis.FunctionResolver.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:53 PM) than earliest output 'C:\github\roslyn\artifacts\bin\Microsoft.CodeAnalysis.FunctionResolver\Debug\net45\Microsoft.CodeAnalysis.ExpressionEvaluator.FunctionResolver.xml' (9/30/2021 12:03:34 PM), not up to date. (Microsoft.CodeAnalysis.FunctionResolver)
9>------ Build started: Project: Microsoft.CodeAnalysis.FunctionResolver, Configuration: Debug Any CPU ------
10>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\BuildBoss\Debug\net472\BuildBoss.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\obj\BuildBoss\Debug\net472\BuildBoss.xml' (9/30/2021 12:03:33 PM), not up to date. (BuildBoss)
10>------ Build started: Project: BuildBoss, Configuration: Debug Any CPU ------
11>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\Microsoft.Build.Tasks.CodeAnalysis\Debug\netcoreapp3.1\Microsoft.Build.Tasks.CodeAnalysis.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:53 PM) than earliest output 'C:\github\roslyn\artifacts\obj\Microsoft.Build.Tasks.CodeAnalysis\Debug\netcoreapp3.1\Microsoft.Build.Tasks.CodeAnalysis.xml' (9/30/2021 12:03:35 PM), not up to date. (Microsoft.Build.Tasks.CodeAnalysis)
11>------ Build started: Project: Microsoft.Build.Tasks.CodeAnalysis, Configuration: Debug Any CPU ------
12>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\Roslyn.Test.Performance.Utilities\Debug\net472\Roslyn.Test.Performance.Utilities.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\bin\Roslyn.Test.Performance.Utilities\Debug\net472\Roslyn.Test.Performance.Utilities.xml' (9/30/2021 12:03:33 PM), not up to date. (Roslyn.Test.Performance.Utilities)
12>------ Build started: Project: Roslyn.Test.Performance.Utilities, Configuration: Debug Any CPU ------
13>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\Roslyn.PerformanceTests\Debug\net472\Roslyn.PerformanceTests.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\obj\Roslyn.PerformanceTests\Debug\net472\Roslyn.PerformanceTests.xml' (9/30/2021 12:03:33 PM), not up to date. (Roslyn.PerformanceTests)
13>------ Build started: Project: Roslyn.PerformanceTests, Configuration: Debug Any CPU ------
14>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\Microsoft.CodeAnalysis.ResultProvider\Debug\netstandard1.3\Microsoft.CodeAnalysis.ResultProvider.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\bin\Microsoft.CodeAnalysis.ResultProvider\Debug\netstandard1.3\Microsoft.CodeAnalysis.ExpressionEvaluator.ResultProvider.xml' (9/30/2021 12:03:34 PM), not up to date. (Microsoft.CodeAnalysis.ResultProvider)
14>------ Build started: Project: Microsoft.CodeAnalysis.ResultProvider, Configuration: Debug Any CPU ------
15>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\ResultProvider.NetFX20\Debug\net20\ResultProvider.NetFX20.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\bin\ResultProvider.NetFX20\Debug\net20\Microsoft.CodeAnalysis.ExpressionEvaluator.ResultProvider.xml' (9/30/2021 12:03:33 PM), not up to date. (ResultProvider.NetFX20)
15>------ Build started: Project: ResultProvider.NetFX20, Configuration: Debug Any CPU ------
16>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\RunTests\Debug\net6.0\RunTests.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\bin\RunTests\Debug\net6.0\RunTests.xml' (9/30/2021 12:03:34 PM), not up to date. (RunTests)
16>------ Build started: Project: RunTests, Configuration: Debug Any CPU ------
17>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\VisualBasicErrorFactsGenerator\x64\Debug\net6.0\VisualBasicErrorFactsGenerator.vbproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\bin\VisualBasicErrorFactsGenerator\x64\Debug\net6.0\VBErrorFactsGenerator.xml' (9/30/2021 12:03:34 PM), not up to date. (VisualBasicErrorFactsGenerator)
17>------ Build started: Project: VisualBasicErrorFactsGenerator, Configuration: Debug x64 ------
18>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\VisualBasicSyntaxGenerator\x64\Debug\net6.0\VisualBasicSyntaxGenerator.vbproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:53 PM) than earliest output 'C:\github\roslyn\artifacts\bin\VisualBasicSyntaxGenerator\x64\Debug\net6.0\VBSyntaxGenerator.xml' (9/30/2021 12:03:34 PM), not up to date. (VisualBasicSyntaxGenerator)
18>------ Build started: Project: VisualBasicSyntaxGenerator, Configuration: Debug x64 ------
19>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\CSharpSyntaxGenerator\Debug\net6.0\CSharpSyntaxGenerator.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:53 PM) than earliest output 'C:\github\roslyn\artifacts\obj\CSharpSyntaxGenerator\Debug\net6.0\CSharpSyntaxGenerator.xml' (9/30/2021 12:03:35 PM), not up to date. (CSharpSyntaxGenerator)
19>------ Build started: Project: CSharpSyntaxGenerator, Configuration: Debug Any CPU ------
20>FastUpToDate: Input 'C:\github\roslyn\artifacts\obj\CSharpErrorFactsGenerator\x64\Debug\net6.0\CSharpErrorFactsGenerator.csproj.BuildWithSkipAnalyzers' is newer (9/30/2021 12:36:52 PM) than earliest output 'C:\github\roslyn\artifacts\bin\CSharpErrorFactsGenerator\x64\Debug\net6.0\CSharpErrorFactsGenerator.xml' (9/30/2021 12:03:34 PM), not up to date. (CSharpErrorFactsGenerator)
20>------ Build started: Project: CSharpErrorFactsGenerator, Configuration: Debug x64 ------
```

So you do effectively a full build every time.

**Comment by vatsalyaagrawal** [comment]
@mavasani please take a look.

**Comment by mavasani** [comment]
@CyrusNajmabadi I was able to repro the issue. The underlying bug was fixed by #56371 and is part of the compiler toolset in the latest VS intpreview build. However, Roslyn.sln doesn't use the compiler toolset that ships in VS for building, but uses a specific compiler toolset NuGet that comes from arcade and the fix is not part of that: https://github.com/dotnet/roslyn/blob/4ba784bd9d6778fbcc99fc234c5946078891736d/eng/Version.Details.xml#L21

The issue should get fixed once we move to building Roslyn.sln with a compiler toolset that has #56371 fix. You can temporarily workaround by doing the following steps:
1. Build Roslyn.sln
2. Open task manager, right click on VBCSCompiler.exe and select `Open file location`
3. Open `Microsoft.Managed.Core.targets` in the opened folder and manually apply the single line fix in #56371

**Comment by mavasani** [comment]
Closing this issue as a duplicate of https://github.com/dotnet/roslyn/issues/56370
